certs/service.go

```diff
@@ -0,0 +1,441 @@
+	ErrKeyBitsValueWrong = errors.New("missing RSA bits for certificate creation")
+
+	// ErrMissingCACertificate indicates missing CA certificate for certificate signing
+	ErrMissingCACertificate = errors.New("missing CA certificate for certificate signing")
+
+	// ErrFailedSerialGeneration failed to generate certificate serial
+	ErrFailedSerialGeneration = errors.New("failed to generate certificate serial")
+
+	// ErrFailedPemKeyWrite indicates problem with writing PEM key
+	ErrFailedPemKeyWrite = errors.New("failed to write PEM key")
+
+	// ErrFailedPemDataWrite failed to write pem data for certificate
+	ErrFailedPemDataWrite = errors.New("failed to write pem data for certificate")
+
+	// ErrPrivateKeyUnsupportedType indicates problem with unsupported  private key type
+	ErrPrivateKeyUnsupportedType = errors.New("private key type is unsupported")
+
+	// ErrPrivateKeyEmpty indicates private key empty
+	ErrPrivateKeyEmpty = errors.New("private key is empty")
+
+	// ErrMissingCertSerial indicates problem with missing certificate serial
+	ErrMissingCertSerial = errors.New("missing cert serial")
+
+	// ErrFailedToRemoveCertFromDB indicates problem in removing cert from db
+	ErrFailedToRemoveCertFromDB = errors.New("failed to remove cert serial from db")
+
+	// ErrFailedToParseCertificate indicates problem parsing certificate
+	ErrFailedToParseCertificate = errors.New("failed to parse x509 certificate")
+
+	// ErrFailedCertDecoding indicates problem with decoding response from PKI service
+	ErrFailedCertDecoding = errors.New("failed to decode response from PKI service")
+)
+
+var _ Service = (*certsService)(nil)
+
+// Service specifies an API that must be fulfilled by the domain service
+// implementation, and all of its decorators (e.g. logging & metrics).
+type Service interface {
+	// IssueCert issues certificate for given thing id if access is granted with token
+	IssueCert(ctx context.Context, token, thingID, daysValid string, keyBits int, keyType string) (Cert, error)
+
+	// ListCertificates lists all certificates issued for given thing
+	ListCertificates(ctx context.Context, token, thingID string, offset, limit uint64) (Page, error)
+
+	// RevokeCert
+	RevokeCert(ctx context.Context, token, thingID, certID string) (Revoke, error)
+}
+
+type Config struct {
+	LogLevel       string
+	ClientTLS      bool
+	CaCerts        string
+	HTTPPort       string
+	ServerCert     string
+	ServerKey      string
+	BaseURL        string
+	ThingsPrefix   string
+	JaegerURL      string
+	AuthnURL       string
+	AuthnTimeout   time.Duration
+	SignTLSCert    tls.Certificate
+	SignX509Cert   *x509.Certificate
+	SignRSABits    int
+	SignHoursValid string
+	PKIHost        string
+	PKIPath        string
+	PKIRole        string
+	PKIToken       string
+}
+
+type certsService struct {
+	auth      mainflux.AuthNServiceClient
+	certsRepo Repository
+	sdk       mfsdk.SDK
+	conf      Config
+	PKIClient *api.Client
```

Comment:
Can we create a wrapper around this client and use it as a lib? Move custom requests and responses to a separate package and import it in certs service? Also, wrap HTTP requests and responses to make it transparent from the service layer.

Reply:
@mteodor Can you, please, take a look at this remark?